diff --git a/CMakeLists.txt b/CMakeLists.txt
index 53816e1..44616ce 100644
--- a/parsers/onnx/CMakeLists.txt
+++ b/parsers/onnx/CMakeLists.txt
@@ -120,6 +120,7 @@ if(WIN32)
     HINTS  ${TENSORRT_ROOT} ${TENSORRT_BUILD} ${CUDA_TOOLKIT_ROOT_DIR}
     PATH_SUFFIXES lib lib64 lib/x64)
 else()
+  set(CMAKE_FIND_LIBRARY_SUFFIXES ".so" ".so.1" ".a") #ugly hack to find libmyelin.so.1 if no symlink
   find_library(TENSORRT_LIBRARY_MYELIN myelin
     HINTS  ${TENSORRT_ROOT} ${TENSORRT_BUILD} ${CUDA_TOOLKIT_ROOT_DIR}
     PATH_SUFFIXES lib lib64 lib/x64)
