diff --git a/parsers/caffe/CMakeLists.txt b/parsers/caffe/CMakeLists.txt
index 74fd358..bfca717 100644
--- a/parsers/caffe/CMakeLists.txt
+++ b/parsers/caffe/CMakeLists.txt
@@ -53,7 +53,7 @@ target_include_directories(${SHARED_TARGET}
 
 set_target_properties(${SHARED_TARGET}
     PROPERTIES
-    CXX_STANDARD 11
+    CXX_STANDARD 14
     CXX_STANDARD_REQUIRED YES
     CXX_EXTENSIONS NO
     ARCHIVE_OUTPUT_DIRECTORY "${TRT_OUT_DIR}"
@@ -103,7 +103,7 @@ target_include_directories(${STATIC_TARGET}
 
 set_target_properties(${STATIC_TARGET}
     PROPERTIES
-    CXX_STANDARD 11
+    CXX_STANDARD 14
     CXX_STANDARD_REQUIRED YES
     CXX_EXTENSIONS NO
     ARCHIVE_OUTPUT_DIRECTORY "${TRT_OUT_DIR}"
