diff --git a/plugin/CMakeLists.txt b/plugin/CMakeLists.txt
index 7711727..a72abd9 100644
--- a/plugin/CMakeLists.txt
+++ b/plugin/CMakeLists.txt
@@ -80,7 +80,7 @@ if(BERT_GENCODES)
     include_directories(bertQKVToContextPlugin/fused_multihead_attention/include bertQKVToContextPlugin/fused_multihead_attention_v2/include)
 endif()
 
-include_directories(common common/kernels ${CMAKE_SOURCE_DIR}/third_party)
+include_directories(common common/kernels ${CMAKE_SOURCE_DIR}/third_party ${CMAKE_SOURCE_DIR}/third_party/cub)
 
 foreach(PLUGIN_ITER ${PLUGIN_LISTS})
     include_directories(${PLUGIN_ITER})
