diff --git a/tools/setup_helpers/cmake.py b/tools/setup_helpers/cmake.py
index 894559ed43..7887147a28 100644
--- a/tools/setup_helpers/cmake.py
+++ b/tools/setup_helpers/cmake.py
@@ -229,6 +229,7 @@ class CMake:
              'CUDA_NVCC_EXECUTABLE',
              'CUDNN_LIBRARY',
              'CUDNN_INCLUDE_DIR',
+             'CAFFE2_LINK_LOCAL_PROTOBUF',
              'EXPERIMENTAL_SINGLE_THREAD_POOL',
              'INSTALL_TEST',
              'JAVA_HOME',
