diff --git a/tools/setup_helpers/cmake.py b/tools/setup_helpers/cmake.py
index f1809552cd..0f3f2eeb89 100644
--- a/tools/setup_helpers/cmake.py
+++ b/tools/setup_helpers/cmake.py
@@ -236,6 +236,7 @@ class CMake:
              'CUDNN_LIBRARY',
              'CUDNN_INCLUDE_DIR',
              'CUDNN_ROOT',
+             'CAFFE2_LINK_LOCAL_PROTOBUF',
              'EXPERIMENTAL_SINGLE_THREAD_POOL',
              'INSTALL_TEST',
              'JAVA_HOME',
