diff --git a/modules/videostab/src/cuda/global_motion.cu b/modules/videostab/src/cuda/global_motion.cu
index 3bc0f0b..f7f91c4 100644
--- a/modules/videostab/src/cuda/global_motion.cu
+++ b/modules/videostab/src/cuda/global_motion.cu
@@ -42,6 +42,7 @@
 #include <thrust/device_ptr.h>
 #include <thrust/remove.h>
 #include <thrust/functional.h>
+#include <thrust/tuple.h>
 #include "opencv2/core/cuda/common.hpp"
 
 namespace cv { namespace cuda { namespace device { namespace globmotion {
@@ -66,7 +67,7 @@ int compactPoints(int N, float *points0, float *points1, const uchar *mask)
     return (int)(thrust::remove_if(thrust::make_zip_iterator(thrust::make_tuple(dpoints0, dpoints1)),
                              thrust::make_zip_iterator(thrust::make_tuple(dpoints0 + N, dpoints1 + N)),
                              dmask, is_zero())
-           - thrust::make_zip_iterator(make_tuple(dpoints0, dpoints1)));
+           - thrust::make_zip_iterator(thrust::make_tuple(dpoints0, dpoints1)));
 }
 
 
