diff --git a/torchvision/extension.py b/torchvision/extension.py
index 265c989a..83e18345 100644
--- a/torchvision/extension.py
+++ b/torchvision/extension.py
@@ -1,8 +1,8 @@
-_HAS_OPS = False
+_HAS_OPS = True
 
 
 def _has_ops():
-    return False
+    return True
 
 
 def _register_extensions():
diff --git a/torchvision/models/detection/_utils.py b/torchvision/models/detection/_utils.py
index db971176..35a50758 100644
--- a/torchvision/models/detection/_utils.py
+++ b/torchvision/models/detection/_utils.py
@@ -53,8 +53,8 @@ class BalancedPositiveNegativeSampler(object):
             num_neg = min(negative.numel(), num_neg)
 
             # randomly select positive and negative examples
-            perm1 = torch.randperm(positive.numel(), device=positive.device)[:num_pos]
-            perm2 = torch.randperm(negative.numel(), device=negative.device)[:num_neg]
+            perm1 = torch.randperm(positive.numel(), device=positive.device, dtype=torch.int64)[:num_pos]
+            perm2 = torch.randperm(negative.numel(), device=negative.device, dtype=torch.int64)[:num_neg]
 
             pos_idx_per_image = positive[perm1]
             neg_idx_per_image = negative[perm2]
